databricks/koalas/frame.py

```diff
@@ -10049,6 +10049,90 @@ def explode(self, column):
         internal = self._internal.copy(spark_frame=sdf, data_spark_columns=data_scols)
         return DataFrame(internal)
 
+    def mad(self, axis=0):
+        """
+        Return the mean absolute deviation of values.
+
+        Parameters
+        ----------
+        axis : {index (0), columns (1)}
+            Axis for the function to be applied on.
+
+        Examples
+        --------
+        >>> df = ks.DataFrame({'a': [1, 2, 3, np.nan], 'b': [0.1, 0.2, 0.3, np.nan]},
+        ...                   columns=['a', 'b'])
+
+        >>> df.mad()
+        a    0.666667
+        b    0.066667
+        Name: 0, dtype: float64
+
+        >>> df.mad(axis=1)
+        0    0.45
+        1    0.90
+        2    1.35
+        3     NaN
+        Name: 0, dtype: float64
+        """
+        from databricks.koalas import Series
+        from databricks.koalas.series import first_series
+
+        axis = validate_axis(axis)
+
+        if axis == 0:
+            exprs = []
+            new_column_labels = []
+            for label in self._internal.column_labels:
+                col_sdf = self._internal.spark_column_for(label)
+                col_type = self._internal.spark_type_for(label)
+
+                if isinstance(col_type, BooleanType):
+                    col_sdf = col_sdf.cast("integer")
+
+                label_str = name_like_string(label)
+
+                col_sdf = F.avg(
+                    F.abs(
+                        col_sdf
+                        - self._sdf.select(F.avg(col_sdf).alias(label_str)).first()[label_str]
+                    )
+                )
+
+                exprs.append(col_sdf.alias(label_str))
+                new_column_labels.append(label)
+
+            sdf = self._sdf.select(exprs)
+
+            with ks.option_context(
+                "compute.default_index_type", "distributed", "compute.max_rows", None
+            ):
+                kdf = DataFrame(sdf)
+                internal = InternalFrame(
+                    kdf._internal.spark_frame,
+                    index_map=kdf._internal.index_map,
+                    column_labels=new_column_labels,
+                    column_label_names=self._internal.column_label_names,
+                )
+
+                return first_series(DataFrame(internal).transpose())
+
+        elif axis == 1:
+            limit = get_option("compute.shortcut_limit")
+            pdf = self.head(limit + 1)._to_internal_pandas()
+            pser = pdf.mad(axis=axis)
+            if len(pdf) <= limit:
+                return Series(pser)
+
+            @pandas_udf(returnType=as_spark_type(pser.dtype.type))
```

Comment:
I took a quick look for `DataFrame.mad` in pandas. Seems it inherits the numpy behaviour from `numpy.mean` which will likely always return a float. I think we can just set the return type as `DoubleType()` and forget about the type infernece.

Reply:
In this case, I think you can also remove the shortcut handling as well.

Reply:
Oh, I agree that type inference is unnecessary! I forgot about it :)